src/Razor/src/Microsoft.VisualStudio.LanguageServer.ContainedLanguage/DefaultFallbackCapabilitiesFilterResolver.cs

```diff
@@ -0,0 +1,242 @@
+                case Methods.TextDocumentReferencesName:
+                    return CheckFindAllReferencesCapabilities;
+                case Methods.TextDocumentRenameName:
+                    return CheckRenameCapabilities;
+                case Methods.TextDocumentSignatureHelpName:
+                    return CheckSignatureHelpCapabilities;
+                case Methods.TextDocumentWillSaveName:
+                    return CheckWillSaveCapabilities;
+                case Methods.TextDocumentWillSaveWaitUntilName:
+                    return CheckWillSaveWaitUntilCapabilities;
+                case Methods.TextDocumentRangeFormattingName:
+                    return CheckRangeFormattingCapabilities;
+                case Methods.WorkspaceSymbolName:
+                    return CheckWorkspaceSymbolCapabilities;
+                case Methods.TextDocumentOnTypeFormattingName:
+                    return CheckOnTypeFormattingCapabilities;
+                case Methods.TextDocumentFormattingName:
+                    return CheckFormattingCapabilities;
+                case Methods.TextDocumentHoverName:
+                    return CheckHoverCapabilities;
+                case Methods.TextDocumentCodeActionName:
+                    return CheckCodeActionCapabilities;
+                case Methods.TextDocumentCodeLensName:
+                    return CheckCodeLensCapabilities;
+                case Methods.TextDocumentCompletionName:
+                    return CheckCompletionCapabilities;
+                case Methods.TextDocumentCompletionResolveName:
+                    return CheckCompletionResolveCapabilities;
+                case Methods.TextDocumentDefinitionName:
+                    return CheckDefinitionCapabilities;
+                case Methods.TextDocumentDocumentHighlightName:
+                    return CheckHighlightCapabilities;
+
+                // VS LSP Expansion capabilities
+                case MSLSPMethods.DocumentReferencesName:
+                    return CheckMSReferencesCapabilities;
+                case MSLSPMethods.ProjectContextsName:
+                    return CheckProjectContextsCapabilities;
+                case MSLSPMethods.TextDocumentCodeActionResolveName:
+                    return CheckCodeActionResolveCapabilities;
+                case MSLSPMethods.OnAutoInsertName:
+                    return CheckOnAutoInsertCapabilities;
+                case MSLSPMethods.DocumentPullDiagnosticName:
+                case MSLSPMethods.WorkspacePullDiagnosticName:
+                    return CheckPullDiagnosticCapabilities;
+
+                default:
+                    return FallbackCheckCapabilties;
+            }
+        }
+
+        private static bool CheckImplementationCapabilities(JToken token)
+        {
+            var serverCapabilities = token.ToObject<ServerCapabilities>();
+
+            return serverCapabilities.ImplementationProvider != null;
```

Comment:
What if the `ServerCapabilities` conversion fails. Would all these benefit from a null conditional operator?

```suggestion
            return serverCapabilities?.ImplementationProvider != null;
```

Reply:
Given the end-to-end I don't think it's possible but given this method is more scoped we can definitely add that restriction.